### PR TITLE
[FEATURE REQUEST] Manage deep links and open the app correctly 

### PR DESCRIPTION
**Pull request by manuelplazaspalacio**
## Related Issues
App: https://github.com/owncloud/android/issues/4181

- [x] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

### CODE REVIEW

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            throw DeepLinkException()
+        }
+        return null
+    }
+
+    data class Params(val uri: URI)
+
+    companion object {
+        const val PATH_SEPARATOR = "/"
+        const val DEEP_LINK_PREVIOUS_PATH_SEGMENT = "f"
```
Comment (jesmrec) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt:43]: `f` is use for oC10 instances, but oCIS is using `s` in the links. Should this be also included?

**Comment by manuelplazaspalacio** [comment]
Finally I used an exiting text inside strings "No account found", to reuse that string. I changed the error message too.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt [hunk 2]
```diff
@@ -63,7 +66,8 @@ class FileOperationsViewModel(
     private val setFilesAsAvailableOfflineUseCase: SetFilesAsAvailableOfflineUseCase,
     private val unsetFilesAsAvailableOfflineUseCase: UnsetFilesAsAvailableOfflineUseCase,
     private val contextProvider: ContextProvider,
-    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider
+    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
+    private val manageDeepLinkUseCase: ManageDeepLinkUseCase,
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt:70]: Add the use cases as parameters before the providers in the ViewModels:

```suggestion
    private val manageDeepLinkUseCase: ManageDeepLinkUseCase,
    private val contextProvider: ContextProvider,
    private val coroutinesDispatcherProvider: CoroutinesDispatcherProvider,
```

File: owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt [hunk 3]
```diff
@@ -95,6 +99,8 @@ class FileOperationsViewModel(
 
     val openDialogs = mutableListOf<FileAlreadyExistsDialog>()
 
+    private val _deepLinkFlow = MutableStateFlow<Event<UIResult<OCFile?>>?>(null)
+    val deepLinkFlow: StateFlow<Event<UIResult<OCFile?>>?> = _deepLinkFlow
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt:103]: Place these just after the other LiveDatas and Flows

File: owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/**
+ * ownCloud Android client application
+ *
+ * @author Manuel Plazas Palacio
+ *
+ * Copyright (C) 2022 ownCloud GmbH.
```
Comment (JuancaG05) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt:6]:
```suggestion
 * Copyright (C) 2023 ownCloud GmbH.
```
😃

File: owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.owncloud.android.domain.exceptions
+
+class DeepLinkException() : Exception(INVALID_FORMAT)
```
Comment (JuancaG05) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt:23]: `()` are not needed if the class doesn't have parameters in the constructor

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        if (pathParts[pathParts.size - 2] != DEEP_LINK_PREVIOUS_PATH_SEGMENT) {
+            throw DeepLinkException()
+        }
+        return null
+    }
+
+    data class Params(val uri: URI)
+
+    companion object {
+        const val PATH_SEPARATOR = "/"
```
Comment (JuancaG05) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt:42]: You can use `com.owncloud.android.domain.files.model.OCFile.Companion.PATH_SEPARATOR` to keep consistency more easily

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
+/**
+ * ownCloud Android client application
+ *
+ * @author Manuel Plazas Palacio
+ *
+ * Copyright (C) 2022 ownCloud GmbH.
```
Comment (JuancaG05) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt:6]:
```suggestion
 * Copyright (C) 2023 ownCloud GmbH.
```
🤠

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some changes requested here @manuelplazaspalacio! Good job! 👍

File: owncloudApp/src/main/java/com/owncloud/android/presentation/authentication/LoginActivity.kt [hunk 8]
```diff
@@ -188,6 +190,24 @@ class LoginActivity : AppCompatActivity(), SslUntrustedCertDialog.OnSslUntrusted
         initLiveDataObservers()
     }
 
+    private fun handleDeepLink() {
+        if (intent.data != null) {
+            authenticationViewModel.launchedFromDeepLink = true
+            if (com.owncloud.android.presentation.authentication.AccountUtils.getAccounts(baseContext).isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (manuelplazaspalacio) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/authentication/LoginActivity.kt:196]: Done

File: owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt (hunk 2 above)
Comment (manuelplazaspalacio) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt:70]: Done

File: owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt (hunk 3 above)
Comment (manuelplazaspalacio) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/files/operations/FileOperationsViewModel.kt:103]: Done

File: owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt (hunk 4 above)
Comment (manuelplazaspalacio) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt:6]: Done

File: owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt (hunk 5 above)
Comment (manuelplazaspalacio) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/exceptions/DeepLinkException.kt:23]: Done

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt (hunk 6 above)
Comment (manuelplazaspalacio) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt:42]: Done

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt (hunk 7 above)
Comment (manuelplazaspalacio) [reply owncloudDomain/src/main/java/com/owncloud/android/domain/files/usecases/ManageDeepLinkUseCase.kt:6]: Done

**Review by JuancaG05 — APPROVED** [review]
LGTM!!

**Pull request merged**